ext/include/opentelemetry/ext/zpages/threadsafe_span_data.h

```diff
@@ -0,0 +1,171 @@
+#pragma once
+
+#include <chrono>
+#include <mutex>
+#include <unordered_map>
+#include <vector>
+
+#include "opentelemetry/core/timestamp.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/sdk/trace/recordable.h"
+#include "opentelemetry/trace/canonical_code.h"
+#include "opentelemetry/trace/span.h"
+#include "opentelemetry/trace/span_id.h"
+#include "opentelemetry/trace/trace_id.h"
+
+using opentelemetry::sdk::trace::AttributeConverter;
+using opentelemetry::sdk::trace::SpanDataAttributeValue;
+namespace trace_api = opentelemetry::trace;
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace ext {
+namespace zpages {
+/**
+ * This class is a threadsafe version of span data used for zpages in OT
+ */
+class ThreadsafeSpanData final : public opentelemetry::sdk::trace::Recordable {
+ public:
+  /**
+   * Get the trace id for this span
+   * @return the trace id for this span
+   */
+  opentelemetry::trace::TraceId GetTraceId() const noexcept {
+    std::unique_lock<std::mutex> lock(mutex_);
```

Comment:
I would change all the getter functions to use [shared_lock](https://en.cppreference.com/w/cpp/thread/shared_lock) for spicy concurrent reads. Those are thread safe and helps speed up the span for whenever getters are called simultaneously

Reply:
If you aren't using shared_lock, I think [lock_guard](https://en.cppreference.com/w/cpp/thread/lock_guard) is better than unique_lock for most cases since it's lighter and simpler. Any reason in particular on why you prefer unique?

Reply:
shared_lock makes sense only if the mutex is shared, this allows the concurrent behavior that you are talking about. I changed them to lock_guard.